### PR TITLE
[RFC] Add ‘eob’ option to fillchars

### PR DESCRIPTION
**Pull request by FriedSock**
This option allows configuring what character is shown on the empty
lines at the end of a buffer, previously hardcoded to ‘~’

Fixes: https://github.com/neovim/neovim/issues/4181

### CODE REVIEW

File: test/functional/options/fillchars_spec.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
+local helpers = require('test.functional.helpers')(after_each)
+local Screen = require('test.functional.ui.screen')
+local clear, execute = helpers.clear, helpers.execute
+
+if helpers.pending_win32(pending) then return end
```
Comment (justinmk) [inline test/functional/options/fillchars_spec.lua]: This should work just fine on Windows.

File: test/functional/options/fillchars_spec.lua [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    clear()
+    screen = Screen.new(25, 5)
+    screen:attach()
+  end)
+
+  after_each(function()
+    screen:detach()
+  end)
+
+  describe('the "eob" option', function()
```
Comment (justinmk) [inline test/functional/options/fillchars_spec.lua]: describe('"eob" flag', function()

**Comment by FriedSock** [comment]
I noticed a mistake in the documentation/implementation whilst doing this PR. The docs for `'fillchars'` state:
```
	Characters to fill the statuslines and vertical separators.
	It is a comma separated list of items:

	  item		default		Used for ~
	  stl:c		' ' or '^'	statusline of the current window
	  stlnc:c	' ' or '='	statusline of the non-current windows
	  vert:c	'|'		vertical separators |:vsplit|
	  fold:c	'-'		filling 'foldtext'
	  diff:c	'-'		deleted lines of the 'diff' option
	  eob:c		'~'		empty lines at the end of a buffer

	*Any one that is omitted will fall back to the default.  For "stl" and
	"stlnc" the space will be used when there is highlighting, '^' or '='
	otherwise.*
```

However, if one of the items is missing, it will not fall back to the default. This behaviour also exists in vim 7.

File: test/functional/options/fillchars_spec.lua [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    it('renders empty lines at the end of the buffer with eob', function()
+      screen:expect([[
+        ^                         |
+        ~                        |
+        ~                        |
+        ~                        |
+                                 |
+      ]])
+      execute('set fillchars+=eob:\\ ')
```
Comment (justinmk) [inline test/functional/options/fillchars_spec.lua]: how about trying some other characters? Including multibyte characters, if those are supported.

Comment (justinmk) [inline test/functional/options/fillchars_spec.lua]: what happens if it's set to empty string?

**Comment by justinmk** [comment]
LGTM after comments. Seems easy enough.

**Comment by jamessan** [comment]
Thanks for working on this @FriedSock.  I would highly recommend submitting this patch upstream, as well.

File: test/functional/options/fillchars_spec.lua (hunk 3 above)
Comment (FriedSock) [inline test/functional/options/fillchars_spec.lua]: multibyte chars aren't supported, neither is setting anything in fillchars to empty string.

Comment (bfredl) [inline test/functional/options/fillchars_spec.lua]: multibyte chars seem to work fine (maybe you meant composing chars won't)

Comment (FriedSock) [inline test/functional/options/fillchars_spec.lua]: Ah yes sorry, I mistakenly tried this with multiple characters. Have added a multibyte character test case now.

**Comment by justinmk** [comment]
- I don't think it works with multibyte characters, actually. E.g., try this: `Ｘ`
    - `:help 'listchars'` but `:help 'fillchars'` suggests only "stl" and "stlnc" flags must be single-byte. 
    - Probably `win_draw_end()` needs to be fixed.
- Need a note in `vim_diff.txt`, see the 'cpoptions' line for example.

File: test/functional/options/fillchars_spec.lua [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        :set fillchars+=eob:\    |
+      ]])
+      execute('set fillchars+=eob:ñ')
+      screen:expect([[
+        ^                         |
+        ñ                        |
+        ñ                        |
+        ñ                        |
+        :set fillchars+=eob:ñ    |
+      ]])
```
Comment (ZyX-I) [inline test/functional/options/fillchars_spec.lua]: I would suggest also checking fullwidth (e.g. U+FF21), composing (both `a`+U+0340 and just U+0340), invalid unicode (`let &fillchars="eob:\x80"`), control characters (`let &fillchars="eob:\n"`), Unicode control characters (`let &fillchars="eob:\u0094"`), just two characters (`ab`). Note: I am not suggesting that all of this must work. Checking errors is also needed. Especially interesting just U+0340: try using `let &fillchars="diff:\u0340"` and running a diff, result is weird.

**Review by ZyX-I — CHANGES_REQUESTED** [review]

File: test/functional/options/fillchars_spec.lua (hunk 4 above)
Comment (ZyX-I) [inline test/functional/options/fillchars_spec.lua]: And missing test for empty fillchars, especially empty after successfully setting to something non-standard.

**Review by ZyX-I — CHANGES_REQUESTED** [review]

**Comment by LeOtaku** [comment]
Any reason this has been delayed for so long? What is preventing this from being pushed forward?

**Comment by KillTheMule** [comment]
The author hasn't been active for some time. Someone needs to pick this up and finish it.

**Comment by bfredl** [comment]
I might be looking into this. I will also consider converting fillchars to schar_T (#7992) representation which maybe makes the code simpler and will allow composing chars as fillchars (mostly because why not), though if this is too messy in a separate PR.

**Comment by bfredl** [comment]
Rebased as #8546

**Comment by justinmk** [comment]
Closed by https://github.com/neovim/neovim/pull/8546
